### PR TITLE
Support collector tags, similar to agent tags

### LINKED ISSUE #1844: Suport for collector level tags
Similar to agent tags https://github.com/jaegertracing/jaeger/issues/833
Related to removing duplicated tags in collector https://github.com/jaegertracing/jaeger/issues/1778

Add the ability to the collector to add a set of tags to incoming spans. 

The agent supports `JAEGER_TAGS` for adding tags, however if we support the same env var in collector the same set of tags could be added twice - although it depends if we remove duplicate tags in #1778.

Use cases:
* add tenant information cc) @radekg

### PR DESCRIPTION
**Pull request by radekg**
## Which problem is this PR solving?

This PR implements the collector Jaeger tags: Resolves #1844.

## Short description of the changes

Adds support for collector `--jaeger.tags` attribute. Reuses Jaeger tags handling code from the agent.

### CODE REVIEW

File: cmd/agent/app/reporter/flags.go [hunk 1]
```diff
@@ -52,45 +52,6 @@ func AddFlags(flags *flag.FlagSet) {
 // InitFromViper initializes Options with properties retrieved from Viper.
 func (b *Options) InitFromViper(v *viper.Viper) *Options {
 	b.ReporterType = Type(v.GetString(reporterType))
-	b.AgentTags = parseAgentTags(v.GetString(agentTags))
+	b.AgentTags = flags.ParseJaegerTags(v.GetString(agentTags))
 	return b
 }
-
-// Parsing logic borrowed from jaegertracing/jaeger-client-go
-func parseAgentTags(agentTags string) map[string]string {
```
Comment (yurishkuro) [inline cmd/agent/app/reporter/flags.go]: Note that this package currently has 100% coverage
```
$ go test -cover ./cmd/agent/app/reporter/
ok  	github.com/jaegertracing/jaeger/cmd/agent/app/reporter	0.069s	coverage: 100.0% of statements
```
which means there is a test covering this function, which should also move to the new location.

File: cmd/collector/app/span_processor.go [hunk 2]
```diff
@@ -159,6 +161,11 @@ func (sp *spanProcessor) enqueueSpan(span *model.Span, originalFormat SpanFormat
 	//add format tag
 	span.Tags = append(span.Tags, model.String("internal.span.format", string(originalFormat)))
 
+	// append the collector tags
+	for k, v := range sp.collectorTags {
```
Comment (yurishkuro) [inline cmd/collector/app/span_processor.go]: add `// TODO add support for deduping tags, https://github.com/jaegertracing/jaeger/issues/1778`

Comment (radekg) [reply cmd/collector/app/span_processor.go]: Done.

File: cmd/agent/app/reporter/flags.go (hunk 1 above)
Comment (radekg) [reply cmd/agent/app/reporter/flags.go]: I am not sure if moving this test out of there is a good idea? It does what it should be doing, parses agent options with tags. I've added additional tests to `cmd/flags`. I could remove `TestBindFlags` test from `cmd/agent/app/reporter/flags_test.go` but I think that would leave the actual option handling without coverage?

Moving the complete test from `cmd/agent/app/reporter/flags_test.go` to `cmd/flags/flags_test.go` does not seem to be an option as is results in circular import.

**Comment by radekg** [comment]
@yurishkuro I'm not sure if the tests are failing because of my changes, first look suggests it's not my changes but I could be wrong.

**Comment by yurishkuro** [comment]
@radekg yes, the failure is due to this change, because in all-in-one the flags from agent and collector are combined.
```
+docker logs 45563e652ba5f0640d4d26904178ab60d5ee9647896330923ead9db0e960e79d
2019/10/13 21:31:09 maxprocs: Leaving GOMAXPROCS=2: CPU quota undefined
flag redefined: jaeger.tags
panic: flag redefined: jaeger.tags
```

**Comment by radekg** [comment]
Not quite sure what's happening now. https://travis-ci.org/jaegertracing/jaeger/builds/597511652?utm_source=github_status&utm_medium=notification shows tests passing but fuzz testing being cancelled.

**Comment by jpkrohling** [comment]
The fuzzing job got cancelled because two previous jobs failed, which I manually restarted (and have now succeeded). I manually triggered the fuzz job as well.

**Comment by radekg** [comment]
Thank you @jpkrohling. Tests are passing now. This feature is complete and ready for review. @yurishkuro, I've renamed the collector flag to `--collector.jaeger.tags`, wasn't initially aware of the clash in `all-in-one` case. Obviously, it would not be possible to share the flag in that case.

**Comment by objectiser** [comment]
Wondering if would be better to call the new flag `collector.tags` and then rename (via deprecation) `jaeger.tags` to `agent.tags`?

**Comment by radekg** [comment]
@objectiser this is a good idea, done in the latest commit.

**Comment by pavolloffay** [comment]
> Wondering if would be better to call the new flag collector.tags and then rename (via deprecation) jaeger.tags to agent.tags?

What is the motivation behind this? Both add tags to process object.

**Comment by radekg** [comment]
Is it reasonable to assume a user might want to add different tags on the agent and others on the collector?

**Comment by objectiser** [comment]
@pavolloffay When used together in the `all-in-one` it just makes it clear that one is being set by the collector and the other when the agent is used. `jaeger.tags` is unclear in this context - and use of `jaeger` in the name is irrelevant anyway, as it is obviously a jaeger parameter.

**Comment by jpkrohling** [comment]
> Is it reasonable to assume a user might want to add certain tags on the agent and others on the collector?

If they are in different processes, yes. If they are in the same process (all-in-one), I don't think it makes much sense to add tags to the agent *and* to the collector.

**Comment by objectiser** [comment]
If it is possible to skip adding the flag to the agent, when running inside `all-in-one`, then that would probably be the best solution.

**Comment by radekg** [comment]
There are 2 options here:

- keep separate flags, maintains predictability in the behavior, agent tags belong to an agent, collector tags to a collector, prevents a potential problem the user would be facing when migrating from `all-in-one` to distributed
- use a single flag for both components, the precedence does not really matter, the complexity for the user arises when moving from `all-in-one` (for example: `agent_tag_1=x,collector_tag_1=y`), where the user might have to split agent and collector related tags, this is inconvenient because it is not possible to copy paste settings (think helm charts, docker compose and so on), another use case here would be having a local dev setup and a production setup, migrating dev setup to prod takes that additional step of having to manually amend the config; from experience, condition based configuration is often a source of problems

It is possible to find out if the flag was defined by using `flag.Lookup`. The question would be which flags take precedence in `all-in-one` but that feels like sprinkling magic instead of having a predictable interface.

The question is: which solution is preferred. My preference would be to keep separate flags.

**Comment by radekg** [comment]
Hi everyone, what's the status on this? It's not clear to me in what direction should this be taken.

**Comment by jpkrohling** [comment]
Sorry, I missed this one. I'd say that each can have its own flag, but the all-in-one should not accept the agent's flag.

**Comment by radekg** [comment]
@jpkrohling `not accept` as in ignore or not even have a flag available.

**Comment by jpkrohling** [comment]
Ideally, we shouldn't even have a flag available, but I don't remember whether we have a flagset per binary or per component, so, not sure it's doable.

**Comment by radekg** [comment]
We could probably do something like this. I could extract this env var handling into a separate package. But this feels really hacky. Is there an existing way of figuring out if we are in `all-in-one` environment?

```diff
diff --git a/cmd/agent/app/reporter/flags.go b/cmd/agent/app/reporter/flags.go
index bd32fb4..6df0e61 100644
--- a/cmd/agent/app/reporter/flags.go
+++ b/cmd/agent/app/reporter/flags.go
@@ -17,6 +17,7 @@ package reporter
 import (
 	"flag"
 	"fmt"
+	"os"

 	"github.com/spf13/viper"
 	"go.uber.org/zap"
@@ -49,7 +50,9 @@ type Options struct {
 func AddFlags(flags *flag.FlagSet) {
 	flags.String(reporterType, string(GRPC), fmt.Sprintf("Reporter type to use e.g. %s, %s", string(GRPC), string(TCHANNEL)))
 	flags.String(agentTagsDeprecated, "", "(deprecated) see --"+agentTags)
-	flags.String(agentTags, "", "One or more tags to be added to the Process tags of all spans passing through this agent. Ex: key1=value1,key2=${envVar:defaultValue}")
+	if _, ok := os.LookupEnv("ALL_IN_ONE_SETUP_CONTEXT"); !ok {
+		flags.String(agentTags, "", "One or more tags to be added to the Process tags of all spans passing through this agent. Ex: key1=value1,key2=${envVar:defaultValue}")
+	}
 }

 // InitFromViper initializes Options with properties retrieved from Viper.
diff --git a/cmd/all-in-one/main.go b/cmd/all-in-one/main.go
index 5ee5c83..a9b71cf 100644
--- a/cmd/all-in-one/main.go
+++ b/cmd/all-in-one/main.go
@@ -154,6 +154,7 @@ by default uses only in-memory database.`,
 	command.AddCommand(env.Command())
 	command.AddCommand(docs.Command(v))

+	os.Setenv("ALL_IN_ONE_SETUP_CONTEXT", "1")
 	config.AddFlags(
 		v,
 		command,
@@ -167,6 +168,7 @@ by default uses only in-memory database.`,
 		queryApp.AddFlags,
 		strategyStoreFactory.AddFlags,
 	)
+	os.Unsetenv("ALL_IN_ONE_SETUP_CONTEXT")

 	if err := command.Execute(); err != nil {
 		fmt.Println(err.Error())
```

**Comment by radekg** [comment]
Nah, if I'm about to add another package then there is no need to do this via env vars. What am I even talking about.

File: cmd/all-in-one/setupcontext/setupcontext.go [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+var isAllInOne bool
+
+// SetAllInOne sets the internal flag to all in one on.
+func SetAllInOne() {
+	isAllInOne = true
+}
+
+// UnsetAllInOne unsets the internal all-in-one flag. Used in tests.
+func UnsetAllInOne(t *testing.T) {
```
Comment (yurishkuro) [inline cmd/all-in-one/setupcontext/setupcontext.go]: please remove the argument, it is not used and will increase the size of the binary.

File: cmd/collector/app/span_processor.go [hunk 4]
```diff
@@ -147,6 +149,13 @@ func (sp *spanProcessor) processItemFromQueue(item *queueItem) {
 	sp.metrics.InQueueLatency.Record(time.Since(item.queuedTime))
 }
 
+func (sp *spanProcessor) addCollectorTags(span *model.Span) {
+	// TODO add support for deduping tags, https://github.com/jaegertracing/jaeger/issues/1778
+	for k, v := range sp.collectorTags {
+		span.Tags = append(span.Tags, model.String(k, v))
```
Comment (yurishkuro) [inline cmd/collector/app/span_processor.go]: iirc the agent adds tags to Span.Process. 

cc @jpkrohling

**Comment by jpkrohling** [comment]
The build is failing with:

``
*** directories without *_test.go files:

./cmd/all-in-one/setupcontext

error: at least one *_test.go file must be in all directories with go files so that they are counted for code coverage

       if no tests are possible for a package (e.g. it only defines types), create empty_test.go
```

Is that new? I don't remember seeing this requirement before...

**Comment by radekg** [comment]
Maybe something to do with code coverage running only on touched paths? Full run of tests covers this. I’ll add a test.

**Comment by yurishkuro** [comment]
It's now new, we had this validation for a long time. Go does not report coverage for packages with no tests, creating misleading numbers. A rudimentary test file would fix it.

**Comment by jpkrohling** [comment]
And directories with `.nocover` would bypass this error, I assume. That's probably why I don't remember seeing that message before.

**Comment by yurishkuro** [comment]
Yes but we should really get rid of those nocover files, since we relaxed 100% coverage requirement.

**Comment by radekg** [comment]
Test added.

File: .travis.yml [hunk 5]
```diff
@@ -36,6 +36,15 @@ matrix:
...
     - HOTROD=true
+  - stage: Fuzz regression
+    dist: bionic
+    go: 1.12.x
+    script: ./scripts/travis/fuzzit.sh local-regression
+  - stage: Fuzz
+    if: branch = master AND type IN (push)
+    dist: bionic
+    go: 1.12.x
+    script: ./scripts/travis/fuzzit.sh fuzzing
```
Comment (yurishkuro) [inline .travis.yml]: These changes don't belong to this PR. Please sync with master

**Comment by radekg** [comment]
ping

**Comment by jpkrohling** [comment]
I'm merging this, as I don't think there's any remaining concern.

**Comment by radekg** [comment]
Nice, thank you @jpkrohling.
